node_modules/gh-users/lib/api.js

```diff
@@ -324,6 +335,243 @@ var updateAdminStatus = module.exports.updateAdminStatus = function(ctx, userId,
+                return callback();
+            }
+        });
+    } else {
+        return callback();
+    }
+};
+
+/**
+ * Import a set of users
+ *
+ * @param  {Number}     app                         The id of the application to import the users into
+ * @param  {String}     authenticationStrategy      The authentication strategy for the user. One of `local` or `shibboleth`
+ * @param  {Boolean}    [forceProfileUpdate]        Whether the user information should be updated, even when other user information is already present
+ * @param  {Object[]}   users                       The users to import
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.err                An error object, if any
+ * @api private
+ */
+var _importUsers = function(app, authenticationStrategy, forceProfileUpdate, users, callback) {
+    if (_.isEmpty(users)) {
+        return callback();
+
+    // Add a progress log statement every 50 imported users
+    } else if (users.length % 50 === 0) {
+        log().info({
+            'authenticationStrategy': authenticationStrategy,
+            'app': app
+        }, 'Importing users from CSV. ' + users.length + ' users left to import');
+    }
+
+    // Import the next user
+    var user = users.pop();
+    _importUser(app, authenticationStrategy, forceProfileUpdate, user, function(err) {
+        // Swallow the error so that 1 bad user account doesn't prevent the others
+        // from being created. The error will have been logged already
+
+        // Recursively import the next user (if any)
+        _importUsers(app, authenticationStrategy, forceProfileUpdate, users, callback);
+    });
+};
+
+/**
+ * Import a user
+ *
+ * @param  {Number}     app                         The id of the application to import the user into
+ * @param  {String}     authenticationStrategy      The authentication strategy for the user. One of `local` or `shibboleth`
+ * @param  {Boolean}    [forceProfileUpdate]        Whether the user information should be updated, even when other user information is already present
+ * @param  {Object}     user                        The user to import
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.err                An error object, if any
+ * @api private
+ */
+var _importUser = function(app, authenticationStrategy, forceProfileUpdate, user, callback) {
+    var userProfile = {
+        'displayName': user[0],
+        'emailPreference': 'immediate'
+    };
+
+    // Shibboleth users don't require an email address. The CSV file returns the empty which would
+    // trigger a constraint violation if we just passed it on. To avoid that, we only set the email
+    // address if one was actually provided
+    if (user[1]) {
+        userProfile.email = user[1];
+    }
+
+    var credentials = {
+        'strategy': authenticationStrategy
+    };
+    var identifier = null;
```

Comment:
I think it's good if our variable nomenclature (down to the DAO) could indicate `externalId`. Ideally we could split it out of the user table as well rather than overloading it, but that can hold off until we introduce another authentication strategy that introduces another `externalId`.